File: src/main/java/retouch/project/careNdShare/service/DonateRequestService.java
package retouch.project.careNdShare.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import retouch.project.careNdShare.dto.DonateRequestDTO;
import retouch.project.careNdShare.dto.DonateRequestResponseDTO;
import retouch.project.careNdShare.entity.DonateItem;
import retouch.project.careNdShare.entity.DonateRequest;
import retouch.project.careNdShare.entity.User;
import retouch.project.careNdShare.entity.DonationStatus;
import retouch.project.careNdShare.entity.DonateRequestStatus;
import retouch.project.careNdShare.repository.DonateItemRepository;
import retouch.project.careNdShare.repository.DonateRequestRepository;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class DonateRequestService {

    @Autowired
    private DonateRequestRepository donateRequestRepository;

    @Autowired
    private DonateItemRepository donateItemRepository;

    @Autowired
    private DonateItemService donateItemService; // Used to claim the item

    @Autowired
    private AuthService authService;

    @Autowired
    private EmailService emailService; // Add this line

    // ✅ Create a new donation request
    @Transactional
    public DonateRequestResponseDTO createRequest(DonateRequestDTO dto) {
        User currentUser = authService.getCurrentUser();

        DonateItem item = donateItemRepository.findById(dto.getDonationId())
                .orElseThrow(() -> new RuntimeException("Donation item not found"));

        if (item.getStatus() != DonationStatus.APPROVED) {
            throw new RuntimeException("This item is not available for request.");
        }

        if (item.getUser().getId().equals(currentUser.getId())) {
            throw new RuntimeException("You cannot request your own donation.");
        }

        DonateRequest request = new DonateRequest();
        request.setDonateItem(item);
        request.setReceiver(currentUser);
        request.setStatus(DonateRequestStatus.PENDING); // Default status

        // ⚠️ CRITICAL FIX: Set the description from DTO
        request.setDescription(dto.getDescription()); // Add this line

        DonateRequest savedRequest = donateRequestRepository.save(request);
        return new DonateRequestResponseDTO(savedRequest);
    }

    // ✅ Get all donation requests (for admin)
    public List<DonateRequestResponseDTO> getAllRequests(DonateRequestStatus status) {
        List<DonateRequest> requests;
        if (status == null) {
            requests = donateRequestRepository.findAll();
        } else {
            requests = donateRequestRepository.findByStatus(status);
        }
        return requests.stream()
                .map(DonateRequestResponseDTO::new)
                .collect(Collectors.toList());
    }

    // ✅ Get all PENDING requests (for admin)
    public List<DonateRequestResponseDTO> getPendingRequests() {
        return donateRequestRepository.findByStatus(DonateRequestStatus.PENDING).stream()
                .map(DonateRequestResponseDTO::new)
                .collect(Collectors.toList());
    }

    // ✅ Get requests made by the currently logged-in user
    public List<DonateRequestResponseDTO> getMyRequests() {
        User currentUser = authService.getCurrentUser();
        return donateRequestRepository.findByReceiverId(currentUser.getId()).stream()
                .map(DonateRequestResponseDTO::new)
                .collect(Collectors.toList());
    }

    // ✅ Approve a request (for admin) - UPDATED VERSION
    @Transactional
    public DonateRequestResponseDTO approveRequest(Long requestId) {
        DonateRequest request = donateRequestRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        if (request.getStatus() != DonateRequestStatus.PENDING) {
            throw new RuntimeException("This request has already been processed.");
        }

        // 1. Approve the request
        request.setStatus(DonateRequestStatus.APPROVED);
        request.setRejectionReason(null);
        DonateRequest savedRequest = donateRequestRepository.save(request);

        // 2. Mark the donation item as CLAIMED
        donateItemService.claimDonation(request.getDonateItem().getId());

        // 3. Reject all other pending requests for this specific item
        List<DonateRequest> otherRequests = donateRequestRepository.findByDonateItemIdAndStatus(
                request.getDonateItem().getId(), DonateRequestStatus.PENDING);

        for (DonateRequest otherReq : otherRequests) {
            otherReq.setStatus(DonateRequestStatus.REJECTED);
            otherReq.setRejectionReason("Item has been claimed by another user.");
            donateRequestRepository.save(otherReq);

            // ✅ ADDED: Send rejection email to the receiver
            try {
                emailService.sendDonationRejectionNotifications(otherReq);
                System.out.println("✅ Rejection email sent for request ID: " + otherReq.getId());
            } catch (Exception e) {
                System.err.println("⚠️ Failed to send rejection email for request ID " + otherReq.getId() + ": " + e.getMessage());
            }
        }

        // 4. Send approval emails to donor and receiver
        try {
            emailService.sendDonationApprovalNotifications(savedRequest);
            System.out.println("✅ Donation approval emails sent successfully!");
        } catch (Exception e) {
            System.err.println("⚠️ Failed to send donation approval emails: " + e.getMessage());
            // Don't throw the exception - allow the approval to succeed even if email fails
            e.printStackTrace();
        }

        return new DonateRequestResponseDTO(savedRequest);
    }

    // ✅ Reject a request (for admin) - UPDATED VERSION
    @Transactional
    public DonateRequestResponseDTO rejectRequest(Long requestId, String reason) {
        DonateRequest request = donateRequestRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        if (request.getStatus() != DonateRequestStatus.PENDING) {
            throw new RuntimeException("This request has already been processed.");
        }

        if (reason == null || reason.isBlank()) {
            throw new RuntimeException("Rejection reason is required");
        }

        request.setStatus(DonateRequestStatus.REJECTED);
        request.setRejectionReason(reason);
        DonateRequest savedRequest = donateRequestRepository.save(request);

        // ✅ ADDED: Send rejection emails
        try {
            emailService.sendDonationRejectionNotifications(savedRequest);
            System.out.println("✅ Rejection emails sent for request ID: " + savedRequest.getId());
        } catch (Exception e) {
            System.err.println("⚠️ Failed to send rejection emails: " + e.getMessage());
        }

        return new DonateRequestResponseDTO(savedRequest);
    }

    // ✅ Get stats for admin dashboard
    public long getRequestCountByStatus(DonateRequestStatus status) {
        if (status == null) return donateRequestRepository.count();
        return donateRequestRepository.countByStatus(status);
    }
}

File: services/donateRequest.service.js
const { DonateRequest, DonateItem, User } = require('../models');

class DonateRequestService {
  // Create donation request
  async createRequest(requestData, userId) {
    try {
      // Check if donation item exists and is approved
      const donateItem = await DonateItem.findById(requestData.donationId);
      if (!donateItem) {
        throw new Error('Donation item not found');
      }

      if (donateItem.status !== 'APPROVED') {
        throw new Error('This donation is not available for requests');
      }

      // Check if user already has a pending/approved request for this donation
      const existingRequest = await DonateRequest.findOne({
        donationId: requestData.donationId,
        receiverUserId: userId,
        status: { $in: ['PENDING', 'APPROVED'] }
      });

      if (existingRequest) {
        throw new Error('You already have a request for this donation');
      }

      const request = new DonateRequest({
        donationId: requestData.donationId,
        receiverUserId: userId,
        description: requestData.description || '',
        quantity: requestData.quantity || 1,
        status: 'PENDING'
      });

      await request.save();
      return request;
    } catch (error) {
      console.error('Error creating donation request:', error);
      throw error;
    }
  }

  // Get user's donation requests
  async getMyRequests(userId) {
    try {
      const requests = await DonateRequest.find({ receiverUserId: userId })
        .populate({ path: 'donationId', populate: { path: 'userId', select: 'firstName lastName email' } })
        .populate('receiverUserId', 'firstName lastName email')
        .sort({ createdAt: -1 });
      
      return requests;
    } catch (error) {
      console.error('Error fetching user requests:', error);
      throw new Error('Failed to fetch donation requests');
    }
  }

  // Get pending requests (for admin)
  async getPendingRequests() {
    return await DonateRequest.find({ status: 'PENDING' })
      .populate({ path: 'donationId', populate: { path: 'userId', select: 'firstName lastName email' } })
      .populate('receiverUserId', 'firstName lastName email')
      .sort({ createdAt: -1 });
  }

  // Get all requests (for admin)
  async getAllRequests(status = null) {
    const filter = status ? { status } : {};

    return await DonateRequest.find(filter)
      .populate({ path: 'donationId', populate: { path: 'userId', select: 'firstName lastName email' } })
      .populate('receiverUserId', 'firstName lastName email')
      .sort({ createdAt: -1 });
  }

  // Approve request
  async approveRequest(id) {
    const request = await DonateRequest.findById(id);
    if (!request) {
      throw new Error('Request not found');
    }

    request.status = 'APPROVED';
    await request.save();

    // Update donation status to CLAIMED
    const donateItem = await DonateItem.findById(request.donationId);
    if (donateItem) {
      donateItem.status = 'CLAIMED';
      await donateItem.save();
    }

    return request;
  }

  // Reject request
  async rejectRequest(id, reason) {
    const request = await DonateRequest.findById(id);
    if (!request) {
      throw new Error('Request not found');
    }

    request.status = 'REJECTED';
    request.rejectionReason = reason;
    await request.save();

    return request;
  }

  // Get request count by status
  async getRequestCountByStatus(status) {
    return await DonateRequest.countDocuments({ status });
  }
}

module.exports = new DonateRequestService();
